### ISSUE TITLE
update `unnecessary_const` to not flag `const`s in case expressions (required for patterns migration)

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#4049]
More context in https://github.com/dart-lang/linter/issues/4044.

TL;DR some fixes to support safe 3.0 migration will get flagged as having `unnecessary_const`s.

```dart
case const [1, 2]:
case const {'k': 'v'}:
case const {1, 2}:
case const Point(1, 2):
```

(but shouldn't.)

**Comment by bwilkerson** [comment]
They definitely shouldn't get flagged in 3.0 code, but it isn't clear to me that the lint should stop flagging those cases in pre-3.0 code. (That might be what you meant, but it wasn't super clear.)
